Parser.php

```diff
@@ -82,8 +82,19 @@ class Parser
     public function __construct(PartFactory $partFactory)
     {
         $this->partFactory = $partFactory;
+        $this->reset();
```

Comment:
Yay, accidental complexity!

So we're rebooting to fix the actual problem, which is reusing the parser when it's not pure. It's also not respecting separation of concerns, we have to know to call the parse method prior to calling any other methods, hilariously this is one of the things I pointed out earlier to someone else, who thought it wasn't that big a deal. =D

In anycase, I believe what should be done is the parse method should return a ParsedMessage, which should have the stateful bits that are stuck on Parser right now, then it's nice and cleanly separated, and the resetting is implicit.